python/taichi/lang/shell.py

```diff
@@ -1,123 +1,14 @@
-        # make IDLE and IPython happy. Hope our users really love them :)
-
-        try:  # IPython / Jupyter?
-            return 'IPython ' + get_ipython().__class__.__name__
-        except:
-            # Note that we can't simply do `'IPython' in sys.modules`,
-            # since it seems `torch` will import IPython on it's own too..
-            if hasattr(__builtins__, '__IPYTHON__'):
-                return ShellType.IPYBASED
-
-        try:
-            if getattr(sys, 'ps1', sys.flags.interactive):
-                return ShellType.NATIVE
-        except:
-            pass
-
-        return ShellType.SCRIPT
-
-    @staticmethod
-    def create_inspector(name):
-        if name is None:
-            # `inspect` for "Python script"
-            import inspect
-            return inspect
-
-        elif name == ShellType.NATIVE:
-            # `dill.source` for "Python native shell"
-            try:
-                import dill
-            except ImportError as e:
-                raise ImportError(
-                    'In order to run Taichi in Python interactive shell, '
-                    'Please execute `python3 -m pip install --user dill`')
-            return dill.source
-
-        elif name.startswith('IPython'):
-            # `IPython.core.oinspect` for "IPython advanced shell"
-            return IPythonInspectorWrapper()
-
-        else:
-            raise RuntimeError(f'Shell type "{name}" not supported')
-
-    def __init__(self):
-        self.name = self.get_shell_name()
-        if self.name is not None:
-            print(f'[Taichi] Interactive shell detected: {self.name}')
-
-        self.inspector = self.create_inspector(self.name)
-
-    def getsource(self, o):
-        return self.inspector.getsource(o)
-
-    def getsourcelines(self, o):
-        return self.inspector.getsourcelines(o)
-
-    def getsourcefile(self, o):
-        return self.inspector.getsourcefile(o)
-
-
-class IPythonInspectorWrapper:
-    """`inspect` module wrapper for IPython / Jupyter notebook"""
-    def __init__(self):
-        pass
-
-    def getsource(self, o):
-        import IPython
-        return IPython.core.oinspect.getsource(o)
-
-    def getsourcelines(self, o):
-        import IPython
-        lineno = IPython.core.oinspect.find_source_lines(o)
-        lines = IPython.core.oinspect.getsource(o).split('\n')
-        return lines, lineno
-
-    def getsourcefile(self, o):
-        import IPython
-        lineno = IPython.core.oinspect.find_source_lines(o)
-        return f'<IPython:{lineno}>'
-
-
-oinspect = ShellInspectorWrapper()
-
-
+try:
+    import sourceinspect as oinspect
+except ImportError:
+    import taichi as ti
+    ti.warn('`sourceinspect` not installed!')
+    ti.warn('This may cause Taichi not functional in IDLE interactive shell, '
+            'Blender scripting module and Python native shell.')
+    ti.warn('Please run `python3 -m pip install sourceinspect` to install.')
+    import inspect as oinspect
```

Comment:
Thanks for adding this cool feature! 🚀  I only have few questions:
1. Why use `ti.warn` instead of `raise ImportError('any msgs...')` or directly printing the error plus the other suggestions?
2. Why add `import inspect as oinspect` in the exception block cuz I don't think this can really help with the error handling? 
3. Shall we add a `else` block to print sth like "Success!" ?

Reply:
> 1. Why use ti.warn instead of raise ImportError('any msgs...') or directly printing the error plus the other suggestions?

In fact I did add `sourceinspect` to hard deps of `taichi` in `setup.py`. **Just in case** that `sourceinspect` didn't installed properly (somehow), users can still play with `taichi` via `inspect`, and a warning message.

> 2. Why add import inspect as oinspect in the exception block cuz I don't think this can really help with the error handling?

The required interface, `oinspect.getsource`, `oinspect.getsourcelines`, is exactly same among `sourceinspect` (ours) and `inspect` (Python builtin). So we can fallback to `inspect` for compatibility.

> 3. No, **ideally users will get `sourceinspect` installed properly thanks to `pip`'s dep resolver**, let's don't annoy them with any message.

Reply:
Ah I see. Thanks for the explanation. No more questions then!